File: python/tests/test_tridiag.py
"""Thomas solver vs dense/banded references, plus input validation."""

from __future__ import annotations

import numpy as np
import pytest
from scipy.linalg import solve_banded  # reference only — never in the pricer path

from localvol import thomas_solve


def _dense(sub, diag, sup):
    n = diag.size
    a = np.diag(diag)
    if n > 1:
        a += np.diag(sub, -1) + np.diag(sup, 1)
    return a


def test_matches_dense_solve_on_random_systems():
    """Property-style: random diagonally-dominant systems of many sizes."""
    rng = np.random.default_rng(123)
    for n in [1, 2, 3, 5, 17, 64, 201]:
        for _ in range(3):
            sub = rng.uniform(-1.0, 1.0, max(n - 1, 0))
            sup = rng.uniform(-1.0, 1.0, max(n - 1, 0))
            diag = rng.uniform(2.5, 4.0, n) * rng.choice([-1.0, 1.0], n)
            rhs = rng.uniform(-5.0, 5.0, n)
            x = thomas_solve(sub, diag, sup, rhs)
            x_ref = np.linalg.solve(_dense(sub, diag, sup), rhs)
            np.testing.assert_allclose(x, x_ref, rtol=1e-12, atol=1e-12)


def test_matches_scipy_banded():
    rng = np.random.default_rng(7)
    n = 120
    sub = rng.normal(size=n - 1)
    sup = rng.normal(size=n - 1)
    diag = 4.0 + rng.uniform(0.0, 1.0, n)
    rhs = rng.normal(size=n)
    ab = np.zeros((3, n))
    ab[0, 1:] = sup
    ab[1, :] = diag
    ab[2, :-1] = sub
    x_ref = solve_banded((1, 1), ab, rhs)
    np.testing.assert_allclose(thomas_solve(sub, diag, sup, rhs), x_ref, rtol=1e-11)


def test_identity_and_size_one():
    np.testing.assert_allclose(
        thomas_solve(np.zeros(2), np.ones(3), np.zeros(2), np.array([1.0, 2.0, 3.0])),
        [1.0, 2.0, 3.0],
    )
    np.testing.assert_allclose(thomas_solve(np.zeros(0), np.array([2.0]), np.zeros(0), np.array([8.0])), [4.0])


@pytest.mark.parametrize(
    "sub,diag,sup,rhs",
    [
        (np.zeros(1), np.ones(3), np.zeros(2), np.ones(3)),  # bad sub length
        (np.zeros(2), np.ones(3), np.zeros(2), np.ones(4)),  # bad rhs length
        (np.zeros(0), np.zeros(0), np.zeros(0), np.zeros(0)),  # empty
    ],
)
def test_shape_validation(sub, diag, sup, rhs):
    with pytest.raises(ValueError):
        thomas_solve(sub, diag, sup, rhs)


def test_rejects_nonfinite_and_zero_pivot():
    with pytest.raises(ValueError, match="non-finite"):
        thomas_solve(np.zeros(1), np.array([1.0, np.nan]), np.zeros(1), np.ones(2))
    with pytest.raises(ValueError, match="pivot"):
        thomas_solve(np.zeros(1), np.array([0.0, 1.0]), np.zeros(1), np.ones(2))
    # elimination-induced zero pivot: [[1,1],[1,1]] is singular
    with pytest.raises(ValueError, match="pivot"):
        thomas_solve(np.array([1.0]), np.array([1.0, 1.0]), np.array([1.0]), np.ones(2))

File: cpp/tests/test_tridiag.cpp
// Thomas solver: agreement with a dense Gaussian-elimination solve on random
// diagonally-dominant tridiagonal systems, plus input validation.

#include <gtest/gtest.h>

#include <cmath>
#include <random>
#include <vector>

#include "localvol/tridiag.hpp"

namespace {

// Dense Gaussian elimination with partial pivoting — an independent check
// (library banded solvers are only allowed inside tests, per the spec).
std::vector<double> dense_solve(std::vector<std::vector<double>> a, std::vector<double> b) {
    const std::size_t n = b.size();
    for (std::size_t col = 0; col < n; ++col) {
        std::size_t piv = col;
        for (std::size_t r = col + 1; r < n; ++r) {
            if (std::abs(a[r][col]) > std::abs(a[piv][col])) piv = r;
        }
        std::swap(a[col], a[piv]);
        std::swap(b[col], b[piv]);
        for (std::size_t r = col + 1; r < n; ++r) {
            const double f = a[r][col] / a[col][col];
            for (std::size_t c = col; c < n; ++c) a[r][c] -= f * a[col][c];
            b[r] -= f * b[col];
        }
    }
    std::vector<double> x(n);
    for (std::size_t i = n; i-- > 0;) {
        double acc = b[i];
        for (std::size_t c = i + 1; c < n; ++c) acc -= a[i][c] * x[c];
        x[i] = acc / a[i][i];
    }
    return x;
}

}  // namespace

TEST(Tridiag, MatchesDenseSolveOnRandomSystems) {
    std::mt19937_64 rng(7);
    std::uniform_real_distribution<double> u(-1.0, 1.0);
    for (int trial = 0; trial < 20; ++trial) {
        const std::size_t n = 2 + static_cast<std::size_t>(trial) % 30;
        std::vector<double> sub(n - 1), diag(n), sup(n - 1), rhs(n);
        for (std::size_t i = 0; i < n - 1; ++i) {
            sub[i] = u(rng);
            sup[i] = u(rng);
        }
        for (std::size_t i = 0; i < n; ++i) {
            // Diagonally dominant: safe for pivot-free Thomas.
            diag[i] = 4.0 + std::abs(u(rng));
            rhs[i] = u(rng);
        }
        std::vector<std::vector<double>> a(n, std::vector<double>(n, 0.0));
        for (std::size_t i = 0; i < n; ++i) {
            a[i][i] = diag[i];
            if (i > 0) a[i][i - 1] = sub[i - 1];
            if (i + 1 < n) a[i][i + 1] = sup[i];
        }
        const std::vector<double> x = localvol::thomas_solve(sub, diag, sup, rhs);
        const std::vector<double> xd = dense_solve(a, rhs);
        for (std::size_t i = 0; i < n; ++i) {
            EXPECT_NEAR(x[i], xd[i], 1e-12) << "trial " << trial << " row " << i;
        }
    }
}

TEST(Tridiag, SolvesSizeOneSystem) {
    const std::vector<double> x = localvol::thomas_solve({}, {2.0}, {}, {5.0});
    ASSERT_EQ(x.size(), 1u);
    EXPECT_DOUBLE_EQ(x[0], 2.5);
}

TEST(Tridiag, RejectsBadInput) {
    EXPECT_THROW(localvol::thomas_solve({}, {}, {}, {}), std::invalid_argument);
    EXPECT_THROW(localvol::thomas_solve({1.0}, {1.0, 1.0, 1.0}, {1.0}, {1.0, 1.0, 1.0}),
                 std::invalid_argument);  // sub/sup length mismatch
    EXPECT_THROW(localvol::thomas_solve({1.0}, {1.0, 1.0}, {1.0}, {1.0}),
                 std::invalid_argument);  // rhs length mismatch
    const double nan = std::nan("");
    EXPECT_THROW(localvol::thomas_solve({nan}, {1.0, 1.0}, {0.0}, {1.0, 1.0}),
                 std::invalid_argument);
    EXPECT_THROW(localvol::thomas_solve({0.0}, {0.0, 1.0}, {0.0}, {1.0, 1.0}),
                 std::invalid_argument);  // zero pivot
}
